proof_of_stake/src/lib.rs

```diff
@@ -0,0 +1,1814 @@
+    fn write_validator_voting_power(
+        &mut self,
+        key: &Self::Address,
+        value: ValidatorVotingPowers,
+    );
+    /// Write PoS bond (validator self-bond or a delegation).
+    fn write_bond(
+        &mut self,
+        key: &BondId<Self::Address>,
+        value: Bonds<Self::TokenAmount>,
+    );
+    /// Write PoS unbond (unbonded tokens from validator self-bond or a
+    /// delegation).
+    fn write_unbond(
+        &mut self,
+        key: &BondId<Self::Address>,
+        value: Unbonds<Self::TokenAmount>,
+    );
+    /// Write PoS validator set (active and inactive).
+    fn write_validator_set(&mut self, value: ValidatorSets<Self::Address>);
+    /// Write PoS total voting power of all validators (active and inactive).
+    fn write_total_voting_power(&mut self, value: TotalVotingPowers);
+
+    /// Delete an emptied PoS bond (validator self-bond or a delegation).
+    fn delete_bond(&mut self, key: &BondId<Self::Address>);
+    /// Delete an emptied PoS unbond (unbonded tokens from validator self-bond
+    /// or a delegation).
+    fn delete_unbond(&mut self, key: &BondId<Self::Address>);
+
+    /// Transfer tokens from the `src` to the `dest`.
+    fn transfer(
+        &mut self,
+        token: &Self::Address,
+        amount: Self::TokenAmount,
+        src: &Self::Address,
+        dest: &Self::Address,
+    );
+
+    /// Attempt to update the given account to become a validator.
+    fn become_validator(
+        &mut self,
+        address: &Self::Address,
+        staking_reward_address: &Self::Address,
+        consensus_key: &Self::PublicKey,
+        current_epoch: impl Into<Epoch>,
+    ) -> Result<(), BecomeValidatorError<Self::Address>> {
+        let current_epoch = current_epoch.into();
+        let params = self.read_pos_params();
+        let mut validator_set = self.read_validator_set();
+        if self.is_validator(address).is_some() {
+            return Err(BecomeValidatorError::AlreadyValidator(
+                address.clone(),
+            ));
+        }
+        let BecomeValidatorData {
+            consensus_key,
+            state,
+        } = become_validator(
+            &params,
+            address,
+            consensus_key,
+            &mut validator_set,
+            current_epoch,
+        );
+        self.write_validator_staking_reward_address(
+            address,
+            staking_reward_address.clone(),
+        );
+        self.write_validator_consensus_key(address, consensus_key);
+        self.write_validator_state(address, state);
+        self.write_validator_set(validator_set);
+        self.write_validator_address_raw_hash(address);
+        Ok(())
+    }
+
+    /// Check if the given address is a validator by checking that it has some
+    /// state.
+    fn is_validator(
+        &mut self,
+        address: &Self::Address,
+    ) -> Option<Epoched<ValidatorState, OffsetPipelineLen>> {
+        self.read_validator_state(address)
+    }
```

Comment:
Nit: Functions of the form is_* should probably return bools. Otherwise, we should call it something like get_validator_state

Reply:
Good point, updated this to return bool in c464d1541d99120bfda0ce9d7865c729191df025